From patchwork Mon Feb 10 06:37:30 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 11372629
Return-Path: <SRS0=Y5yW=36=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 41EB592A
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 10 Feb 2020 06:37:46 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 203C222464
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 10 Feb 2020 06:37:45 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=gmail.com header.i=@gmail.com
 header.b="Mxve/0KY"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726376AbgBJGhm (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 10 Feb 2020 01:37:42 -0500
Received: from mail-oi1-f195.google.com ([209.85.167.195]:33227 "EHLO
        mail-oi1-f195.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726061AbgBJGhm (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 10 Feb 2020 01:37:42 -0500
Received: by mail-oi1-f195.google.com with SMTP id q81so8234585oig.0;
        Sun, 09 Feb 2020 22:37:42 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=mime-version:from:date:message-id:subject:to:cc;
        bh=sKP+BJ8wBjzEgCx/CYPr7X88CtkSZccKI/5Y02IAgCw=;
        b=Mxve/0KYXH5nHyM1kFSEhNJmYGTlousTYlbYqS1jOPMFF5Veo/9Y2Jz1JOPiMBxd9F
         OYJp9/qUWlj1VlP5SVdNPe3cDJp8BRmqXRUc+QirNj1+jVPKR/LaoO0FKG8JjiD+U8C2
         t/V0PS0BVV4yqtu836kFnbTc2GIylo8fX1pZijFgOvBNGt69qJjka+wCv4XH7ngpfNbA
         7qKphDgW2Ylk7KYUJxic314TvJcqfmzbmHImnajj6+LLQpl2XF7ceLHcpuCVjocUpF4x
         qgjSpPAbumAqrOU5NtK5x7k+FtzTzhUMguu9pukthdhin5KYXOcjaz9I4j09j36Svtyq
         uQqg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:mime-version:from:date:message-id:subject:to:cc;
        bh=sKP+BJ8wBjzEgCx/CYPr7X88CtkSZccKI/5Y02IAgCw=;
        b=q4I+u0kM7VPW0F+4tpv66M4fq4iF0jK6Ydp/dtkfjRSpLWEWcc3TdfQ9kmolDqM1/D
         D97b/isQ41OQB2X5C5RorWNYAz3ySeURpy+XYhWwP487JQLIDpx2UwDyDpmoyj6rOkFB
         6j57CAzBvXEIYxhy+D2RyqfaKJbdZwmPaQVkPIXCBV4bsi9Wc7BgcQ+CcxtZ1iuqvFpl
         KedUnjmnEH+ytrSadgrZgKJaSJ0J6EeA1VXVq7A92oXR7AiYzakGAu/AHHRVYZ5CjPX2
         Mn5DQQckPwfmYzf3KAtw1KCOu8zxvoCnAZSh0fFgectpeGoSj9bmUDdLWPwScPIAflMp
         Am+w==
X-Gm-Message-State: APjAAAWNcM3DM1QK0hKfE9rHA7HoHI8x+EkTfWWxhRR92+d9vHnneGK0
        cEHat6utClocboYd7S5vPPpPCtw3HxOgSb/P6ud5mKYtyP/Xpw==
X-Google-Smtp-Source: 
 APXvYqz1VjnaGereiNdTBdBdTbpaWc/Yme4gnjBca8gp58S465ODJyip6auKSicR/eGvEjwcpEpXy4CtvheHR9SeKG0=
X-Received: by 2002:aca:f305:: with SMTP id r5mr10049656oih.174.1581316661703;
 Sun, 09 Feb 2020 22:37:41 -0800 (PST)
MIME-Version: 1.0
From: Wanpeng Li <kernellwp@gmail.com>
Date: Mon, 10 Feb 2020 14:37:30 +0800
Message-ID: 
 <CANRm+Cxd55Sqi4anpXD_Urmx8BV=R9ZDUwejChJHLBsZeGoWbw@mail.gmail.com>
Subject: [PATCH v2 1/2] KVM: Introduce pv check helpers
To: LKML <linux-kernel@vger.kernel.org>, kvm <kvm@vger.kernel.org>
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

Introduce some pv check helpers for consistency.

Suggested-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kernel/kvm.c | 34 ++++++++++++++++++++++++----------
 1 file changed, 24 insertions(+), 10 deletions(-)

--
2.7.4

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index d817f25..76ea8c4 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -425,7 +425,27 @@ static void __init sev_map_percpu_data(void)
     }
 }

+static bool pv_tlb_flush_supported(void)
+{
+    return (kvm_para_has_feature(KVM_FEATURE_PV_TLB_FLUSH) &&
+        !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
+        kvm_para_has_feature(KVM_FEATURE_STEAL_TIME));
+}
+
 #ifdef CONFIG_SMP
+
+static bool pv_ipi_supported(void)
+{
+    return kvm_para_has_feature(KVM_FEATURE_PV_SEND_IPI);
+}
+
+static bool pv_sched_yield_supported(void)
+{
+    return (kvm_para_has_feature(KVM_FEATURE_PV_SCHED_YIELD) &&
+        !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
+        kvm_para_has_feature(KVM_FEATURE_STEAL_TIME));
+}
+
 #define KVM_IPI_CLUSTER_SIZE    (2 * BITS_PER_LONG)

 static void __send_ipi_mask(const struct cpumask *mask, int vector)
@@ -619,9 +639,7 @@ static void __init kvm_guest_init(void)
         pv_ops.time.steal_clock = kvm_steal_clock;
     }

-    if (kvm_para_has_feature(KVM_FEATURE_PV_TLB_FLUSH) &&
-        !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
-        kvm_para_has_feature(KVM_FEATURE_STEAL_TIME)) {
+    if (pv_tlb_flush_supported()) {
         pv_ops.mmu.flush_tlb_others = kvm_flush_tlb_others;
         pv_ops.mmu.tlb_remove_table = tlb_remove_table;
     }
@@ -632,9 +650,7 @@ static void __init kvm_guest_init(void)
 #ifdef CONFIG_SMP
     smp_ops.smp_prepare_cpus = kvm_smp_prepare_cpus;
     smp_ops.smp_prepare_boot_cpu = kvm_smp_prepare_boot_cpu;
-    if (kvm_para_has_feature(KVM_FEATURE_PV_SCHED_YIELD) &&
-        !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
-        kvm_para_has_feature(KVM_FEATURE_STEAL_TIME)) {
+    if (pv_sched_yield_supported()) {
         smp_ops.send_call_func_ipi = kvm_smp_send_call_func_ipi;
         pr_info("KVM setup pv sched yield\n");
     }
@@ -700,7 +716,7 @@ static uint32_t __init kvm_detect(void)
 static void __init kvm_apic_init(void)
 {
 #if defined(CONFIG_SMP)
-    if (kvm_para_has_feature(KVM_FEATURE_PV_SEND_IPI))
+    if (pv_ipi_supported())
         kvm_setup_pv_ipi();
 #endif
 }
@@ -739,9 +755,7 @@ static __init int kvm_setup_pv_tlb_flush(void)
     if (!kvm_para_available() || nopv)
         return 0;

-    if (kvm_para_has_feature(KVM_FEATURE_PV_TLB_FLUSH) &&
-        !kvm_para_has_hint(KVM_HINTS_REALTIME) &&
-        kvm_para_has_feature(KVM_FEATURE_STEAL_TIME)) {
+    if (pv_tlb_flush_supported()) {
         for_each_possible_cpu(cpu) {
             zalloc_cpumask_var_node(per_cpu_ptr(&__pv_tlb_mask, cpu),
                 GFP_KERNEL, cpu_to_node(cpu));

From patchwork Mon Feb 10 06:38:59 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 11372645
Return-Path: <SRS0=Y5yW=36=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8A246921
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 10 Feb 2020 06:39:14 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 5F48E20661
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 10 Feb 2020 06:39:14 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=gmail.com header.i=@gmail.com
 header.b="DgaeFq+u"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727429AbgBJGjL (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 10 Feb 2020 01:39:11 -0500
Received: from mail-ot1-f68.google.com ([209.85.210.68]:42750 "EHLO
        mail-ot1-f68.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726118AbgBJGjL (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 10 Feb 2020 01:39:11 -0500
Received: by mail-ot1-f68.google.com with SMTP id 66so5244190otd.9;
        Sun, 09 Feb 2020 22:39:10 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=mime-version:from:date:message-id:subject:to:cc;
        bh=HlsjcWP9mUkL0ZSkAPaWKXT3aO8hLJuD8G7R6rOrkyo=;
        b=DgaeFq+u997sASMaYeP69Cfm3FrtMU5HvAXKfwLQv2kd4NSHBGEsiylOn3nWsAKX6/
         v/YruLjveEu+JFC5cPxCiHKZHhNr9or+BMP6TrMsSy9el69bBlfmWj0R4YeQdL6G/Ng5
         KE9QCKR6l2tgmIi3/wof+1Zmhdh/OXU4ctsO5HCplwR9OPCLL4+fL8R8qOmFNe3dMFgb
         nE1LB+WyYnFXd96gdNIiatiQGPldn3tJE7ozzb6w5l9xzzADvshy1zNPXEKT7GIsUiZH
         HficcbrdUpOEDAtqCwfhZmFyeQXELuKLp0Y2yEx2cVVRH5xQGj9SLvvNehR3kyFCAH6P
         8MIA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:mime-version:from:date:message-id:subject:to:cc;
        bh=HlsjcWP9mUkL0ZSkAPaWKXT3aO8hLJuD8G7R6rOrkyo=;
        b=BkBnxVme5SUNhH1l9eYag3zUYRqnmwwOgPyaNDT5yY0QKbLHeaS58hN3ebezNTTQ++
         uAd7s4ET3MGEqfhG036+tqKAyphjuwmDzzKx7FW1B/LwA+qv6/RzpVte8IbMhBf96Qe3
         d5r/UIuV5gYAVsBr6/+Qpzis6tW0Z0nipPaz6KryE7SQ3K6l4qRMT06v6FQaQ4ULqnI7
         jwqBQXQYHD2evnjTUF9q5NoqXFYgaaElXGBS48JM8F46jVn0U1Ojj7Kj6CosxhVKPjqc
         0KJtvrV9xFelgVonDTbfxTHxppdmrA+UJZroIPYApjQgMovrsl5RuzS7KqUEEvxHiR9M
         7PvQ==
X-Gm-Message-State: APjAAAXLlLO9Kh9Unf2a+WkgRt86OAet0+X+r6vwGyrJWp/RXrsY6vPO
        x87rXgy0HGGcBPEBdzQxSqPP3JAZ4oVEPKRFlWFx4ic1eYPp/g==
X-Google-Smtp-Source: 
 APXvYqzPo25A9j9ckAWsBjnev3+3TQTWLXQQtFc33/AoSq2bgjPdrjBd9Wtyam+cPp8vkJf3sQE36TRwavSyrCwFtzY=
X-Received: by 2002:a9d:7ccd:: with SMTP id r13mr19722otn.56.1581316750332;
 Sun, 09 Feb 2020 22:39:10 -0800 (PST)
MIME-Version: 1.0
From: Wanpeng Li <kernellwp@gmail.com>
Date: Mon, 10 Feb 2020 14:38:59 +0800
Message-ID: 
 <CANRm+CxGOeGQ0vV9ueBgjUDvkzH29EQWLe4GQGDvOhm3idM6NQ@mail.gmail.com>
Subject: [PATCH v2 2/2] KVM: Pre-allocate 1 cpumask variable per cpu for both
 pv tlb and pv ipis
To: LKML <linux-kernel@vger.kernel.org>, kvm <kvm@vger.kernel.org>
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

Nick Desaulniers Reported:

  When building with:
  $ make CC=clang arch/x86/ CFLAGS=-Wframe-larger-than=1000
  The following warning is observed:
  arch/x86/kernel/kvm.c:494:13: warning: stack frame size of 1064 bytes in
  function 'kvm_send_ipi_mask_allbutself' [-Wframe-larger-than=]
  static void kvm_send_ipi_mask_allbutself(const struct cpumask *mask, int
  vector)
              ^
  Debugging with:
  https://github.com/ClangBuiltLinux/frame-larger-than
  via:
  $ python3 frame_larger_than.py arch/x86/kernel/kvm.o \
    kvm_send_ipi_mask_allbutself
  points to the stack allocated `struct cpumask newmask` in
  `kvm_send_ipi_mask_allbutself`. The size of a `struct cpumask` is
  potentially large, as it's CONFIG_NR_CPUS divided by BITS_PER_LONG for
  the target architecture. CONFIG_NR_CPUS for X86_64 can be as high as
  8192, making a single instance of a `struct cpumask` 1024 B.

This patch fixes it by pre-allocate 1 cpumask variable per cpu and use it for
both pv tlb and pv ipis..

Reported-by: Nick Desaulniers <ndesaulniers@google.com>
Acked-by: Nick Desaulniers <ndesaulniers@google.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: Nick Desaulniers <ndesaulniers@google.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
v1 -> v2:
 * remove '!alloc' check
 * use new pv check helpers

 arch/x86/kernel/kvm.c | 33 +++++++++++++++++++++------------
 1 file changed, 21 insertions(+), 12 deletions(-)

 }

@@ -595,7 +597,6 @@ static void __init kvm_apf_trap_init(void)
     update_intr_gate(X86_TRAP_PF, async_page_fault);
 }

-static DEFINE_PER_CPU(cpumask_var_t, __pv_tlb_mask);

 static void kvm_flush_tlb_others(const struct cpumask *cpumask,
             const struct flush_tlb_info *info)
@@ -603,7 +604,7 @@ static void kvm_flush_tlb_others(const struct
cpumask *cpumask,
     u8 state;
     int cpu;
     struct kvm_steal_time *src;
-    struct cpumask *flushmask = this_cpu_cpumask_var_ptr(__pv_tlb_mask);
+    struct cpumask *flushmask = this_cpu_cpumask_var_ptr(__pv_cpu_mask);

     cpumask_copy(flushmask, cpumask);
     /*
@@ -642,6 +643,7 @@ static void __init kvm_guest_init(void)
     if (pv_tlb_flush_supported()) {
         pv_ops.mmu.flush_tlb_others = kvm_flush_tlb_others;
         pv_ops.mmu.tlb_remove_table = tlb_remove_table;
+        pr_info("KVM setup pv remote TLB flush\n");
     }

     if (kvm_para_has_feature(KVM_FEATURE_PV_EOI))
@@ -748,24 +750,31 @@ static __init int activate_jump_labels(void)
 }
 arch_initcall(activate_jump_labels);

-static __init int kvm_setup_pv_tlb_flush(void)
+static __init int kvm_alloc_cpumask(void)
 {
     int cpu;
+    bool alloc = false;

     if (!kvm_para_available() || nopv)
         return 0;

-    if (pv_tlb_flush_supported()) {
+    if (pv_tlb_flush_supported())
+        alloc = true;
+
+#if defined(CONFIG_SMP)
+    if (pv_ipi_supported())
+        alloc = true;
+#endif
+
+    if (alloc)
         for_each_possible_cpu(cpu) {
-            zalloc_cpumask_var_node(per_cpu_ptr(&__pv_tlb_mask, cpu),
+            zalloc_cpumask_var_node(per_cpu_ptr(&__pv_cpu_mask, cpu),
                 GFP_KERNEL, cpu_to_node(cpu));
         }
-        pr_info("KVM setup pv remote TLB flush\n");
-    }

     return 0;
 }
-arch_initcall(kvm_setup_pv_tlb_flush);
+arch_initcall(kvm_alloc_cpumask);

 #ifdef CONFIG_PARAVIRT_SPINLOCKS

--
2.7.4

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 76ea8c4..377b224 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -432,6 +432,8 @@ static bool pv_tlb_flush_supported(void)
         kvm_para_has_feature(KVM_FEATURE_STEAL_TIME));
 }

+static DEFINE_PER_CPU(cpumask_var_t, __pv_cpu_mask);
+
 #ifdef CONFIG_SMP

 static bool pv_ipi_supported(void)
@@ -510,12 +512,12 @@ static void kvm_send_ipi_mask(const struct
cpumask *mask, int vector)
 static void kvm_send_ipi_mask_allbutself(const struct cpumask *mask,
int vector)
 {
     unsigned int this_cpu = smp_processor_id();
-    struct cpumask new_mask;
+    struct cpumask *new_mask = this_cpu_cpumask_var_ptr(__pv_cpu_mask);
     const struct cpumask *local_mask;

-    cpumask_copy(&new_mask, mask);
-    cpumask_clear_cpu(this_cpu, &new_mask);
-    local_mask = &new_mask;
+    cpumask_copy(new_mask, mask);
+    cpumask_clear_cpu(this_cpu, new_mask);
+    local_mask = new_mask;
     __send_ipi_mask(local_mask, vector);
